experiments/magrank.py

```diff
@@ -0,0 +1,182 @@
+import streamlit as st
+from .base import Experiment
+from data_tools import (
+    vectorize_text,
+    preprocess_text_col,
+    add_language_feature,
+    add_author_prominence_feature,
+)
+from distribution_study import show_distribution
+from sklearn.linear_model import LinearRegression, Lasso, Ridge
+import numpy as np
+import pandas as pd
+from sklearn.metrics import r2_score
+from sklearn.svm import LinearSVC
+from sklearn.naive_bayes import MultinomialNB
+from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
+from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
+from sklearn.preprocessing import FunctionTransformer, OneHotEncoder, LabelEncoder
+from sklearn.pipeline import Pipeline, FeatureUnion
+from .sklearn_utils import ClfSwitcher
+from sklearn.model_selection import GridSearchCV
+from xgboost import XGBRegressor
+
+
+@st.cache(allow_output_mutation = True)
+def hyperparameter_tuning(pipeline, parameters):
+    return GridSearchCV(pipeline, parameters, cv=5, n_jobs=12, return_train_score=False, verbose=3)
+
+select_columns = lambda features: FunctionTransformer(lambda x: x[features], validate=False)
+
+def print_model_results(Y_pred, Y_test):
+    mean_absolute_error = np.abs(Y_pred - Y_test).mean()
+    mean_percentage_error = (np.abs(Y_pred - Y_test) / Y_test).mean() * 100
+    score = r2_score(Y_test, Y_pred)
+
+    st.markdown(
+        """
+        #### Model results:
+        - Mean absolute error: {:,.0f}
+        - Mean percentage error: {:,.2f}%
+        - R2 Score: {}    
+    """.format(
+            mean_absolute_error, mean_percentage_error, r2_score(Y_test, Y_pred)
+        )
+    )
+
+    return score
+
+
+class MagRankExperiment(Experiment):
+    def __init__(self, pandas_df):
+
+        # Drop non-numeric page features and add PageCount feature
```

Comment:
This is not in use atm. I'll move it to the derived `feature_selection` in a follow-up PR